gcsfs/core.py

```diff
@@ -398,65 +562,77 @@ def rmdir(self, bucket):
             for v in self.dirs[''][:]:
                 if v['name'] == bucket:
                     self.dirs[''].remove(v)
-        self.dirs.pop(bucket, None)
-
-    def invalidate_cache(self, bucket=None):
-        """
-        Mark files cache as dirty, so that it is reloaded on next use.
-
-        Parameters
-        ----------
-        bucket: string or None
-            If None, clear all files cached; if a string, clear the files
-            corresponding to that bucket.
-        """
-        if bucket in {'/', '', None}:
-            self.dirs.clear()
-        else:
-            self.dirs.pop(bucket, None)
+        self.invalidate_cache(bucket)
 
     def ls(self, path, detail=False):
-        if path in ['', '/']:
-            out = self._list_buckets()
-        else:
-            bucket, prefix = split_path(path)
-            path = '/'.join([bucket, prefix])
-            files = self._list_bucket(bucket)
-            seek, l, bit = (path, len(path), '') if path.endswith('/') else (
-                path+'/', len(path)+1, '/')
-            out = []
-            for f in files:
-                if (f['name'].startswith(seek) and '/' not in f['name'][l:] or
-                        f['name'] == path):
-                    out.append(f)
-                elif f['name'].startswith(seek) and '/' in f['name'][l:]:
-                    directory = {
-                        'bucket': bucket, 'kind': 'storage#object',
-                        'size': 0, 'storageClass': 'DIRECTORY',
-                        'name': path+bit+f['name'][l:].split('/', 1)[0]+'/'}
-                    if directory not in out:
-                        out.append(directory)
-        if detail:
-            return out
+        """List objects under the given '/{bucket}/{prefix} path."""
+        if path in ['/', '']:
+            out = self.buckets
         else:
-            return [f['name'] for f in out]
+            if not path.endswith("/"):
+                path = path + "/"
```

Comment:
Is it not in general reasonable to do `ls` on a pull key name, and expect to get the same information as `info` on that file would bring?